### PR TITLE
feat: add auto-focussing via URLSearchParam to content-manager

### PR DESCRIPTION
**Pull request by joshuaellis**
### What does it do?

* Adds new `useFocusInputField` hook exposed via the `helper-plugin`
* Utilises new `useFocusInputField` hook to auto-focus in the content-manager in `GenericInputs`
* Utilises new `useFocusInputField` hook to auto-focus relation fields in the content-manager
* Tweaks `WYSIWG` to be able to auto-focus 
* Updates custom-fields to accept and utilise their forwarded ref
* Automatically opens `RepeatableComponent` if the field we want is located inside

### Why is it needed?

* Makes visual editing better

### How to test it?

* Automated testing
OR
* Go to a kitchen-sink entry
* add `field=<name>` as a query param e.g. `short_text` or `timestamp` or `custom_field`
* to test dynamic zones you need to have them created and saved
* then add their "path" e.g. `cats.0.categories`
* to test repeatable components you need to have them created and saved
* then add their "path" e.g. `repeatable_compo.1. name`

### Related issue(s)/PR(s)

* resolves CONTENT-1590
* requires https://github.com/strapi/design-system/pull/1268 to be merged and the CMS updated

### CODE REVIEW

File: packages/core/admin/admin/src/content-manager/components/RepeatableComponent/index.js [hunk 1]
```diff
@@ -50,6 +50,37 @@ const RepeatableComponent = ({
...
+   * Get the temp key of the component that has the field that is currently focussed
+   * as defined by the `field` query param. We can then force this specific component
+   * to be in it's "open" state.
+   */
+  const componentTmpKeyWithFocussedField = useMemo(() => {
+    if (search.has('field')) {
+      const field = search.get('field');
+
+      // eslint-disable-next-line no-unused-vars
+      let [_, path] = field.split(`${name}.`);
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RepeatableComponent/index.js:65]:
```suggestion
      let [, path] = field.split(`${name}.`);
```

should work too without eslint ignore :)

File: packages/core/helper-plugin/src/components/GenericInput.jsx [hunk 2]
```diff
@@ -441,13 +451,18 @@ const GenericInput = ({
     }
     default: {
       return (
```
Comment (gu-stav) [inline packages/core/helper-plugin/src/components/GenericInput.jsx:453]: What do you think about adding a comment here, this branch renders "not supported"? Previously the name of the component made that clear for me...

File: packages/core/helper-plugin/src/components/tests/GenericInput.test.jsx [hunk 3]
```diff
@@ -231,4 +228,35 @@ describe('GenericInput', () => {
       });
     });
   });
+
+  describe('auto-focussing', () => {
```
Comment (gu-stav) [inline packages/core/helper-plugin/src/components/tests/GenericInput.test.jsx:232]: Very cool test!

**Review by gu-stav — COMMENTED** [review]
Very nice hook and refactoring! 🚀 Could you maybe document a few `field` params that might work to be able to do QA?

File: packages/core/helper-plugin/src/components/GenericInput.jsx (hunk 2 above)
Comment (joshuaellis) [reply packages/core/helper-plugin/src/components/GenericInput.jsx:453]: ✅

**Comment by joshuaellis** [comment]
I've updated the description for the PR if you want to do some manual QA.

File: packages/core/admin/admin/src/content-manager/components/RepeatableComponent/index.js (hunk 1 above)
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/RepeatableComponent/index.js]: You can also remove the eslint ignore comment now :)

**Comment by gu-stav** [comment]
I've done some manual QA:

- [ ] couldn't find a way to set the focus into a component-field that is part of a dynamic zone
- [ ] media fields don't receive a focus (but I am not sure inline editing works with media fields yet?)

All other fields work really well 🚀

**Comment by joshuaellis** [comment]
>  media fields don't receive a focus (but I am not sure inline editing works with media fields yet?)

the field receives focus, the issue is the field imo. You either have it filled or not filled, in the case of not filled the button to add elements is nested deeply inside the field but if it's filled the first action is the one that gets focus (when you tab through the page) this inconsistency is difficult to deal with to be honest.

IMHO it's a design issue that needs a rethink, e.g. could the whole field always be "seen" as a button so you can focus the whole thing and pressing enter lets you change the image etc.

So for now, we focus the field and the first time you press `tab` will pass you to the correct piece – but also, I don't think it's part of visual editing yet because it's not a serialisable primitive (i'd have to check this with Alex)

>  couldn't find a way to set the focus into a component-field that is part of a dynamic zone

Really? I've just tried it and it works, this is my path `dynamiczone.0.repeat_relation.0.categories` where I added a repeatable component to an already existing dynamic zone component – is this what you mean?

**Review by jhoward1994 — DISMISSED** [review]
Very cool, this works well for me!

**Comment by joshuaellis** [comment]
sorry @jhoward1994 i had to rebase 🙏🏼

**Review by gu-stav — APPROVED** [review]

**Pull request merged**